pkg/reconciler/taskrun/resources/apply.go

```diff
@@ -144,6 +144,40 @@ func ApplyWorkspaces(spec *v1beta1.TaskSpec, declarations []v1beta1.WorkspaceDec
 	return ApplyReplacements(spec, stringReplacements, map[string][]string{})
 }
 
+// applyWorkspaceMountPath accepts a workspace path variable of the form $(workspaces.foo.path) and replaces
+// it in the fields of the TaskSpec. A new updated TaskSpec is returned. Steps or Sidecars in the TaskSpec
+// that override the mountPath will receive that mountPath in place of the variable's value. Other Steps and
+// Sidecars will see either the workspace's declared mountPath or the default of /workspaces/<name>.
+func applyWorkspaceMountPath(variable string, spec *v1beta1.TaskSpec, declaration v1beta1.WorkspaceDeclaration) *v1beta1.TaskSpec {
+	stringReplacements := map[string]string{variable: ""}
+	emptyArrayReplacements := map[string][]string{}
+	defaultMountPath := declaration.GetMountPath()
+	// Replace instances of the workspace path variable that are overridden per-Step
+	for i := range spec.Steps {
+		step := &spec.Steps[i]
+		for _, usage := range step.Workspaces {
+			if usage.Name == declaration.Name && usage.MountPath != "" {
+				stringReplacements[variable] = usage.MountPath
+				v1beta1.ApplyStepReplacements(step, stringReplacements, emptyArrayReplacements)
+			}
+		}
+	}
+	// Replace instances of the workspace path variable that are overridden per-Sidecar
+	for i := range spec.Sidecars {
+		sidecar := &spec.Sidecars[i]
+		for _, usage := range sidecar.Workspaces {
+			if usage.Name == declaration.Name && usage.MountPath != "" {
+				stringReplacements[variable] = usage.MountPath
+				v1beta1.ApplySidecarReplacements(sidecar, stringReplacements, emptyArrayReplacements)
+			}
+		}
+	}
```

Comment:
just nitpicking and i dont have a clear alternative suggestion, but it feels like some of this logic belongs to how "workspaces" work and i'd almost expect to see it in the workspaces package

(i.e. which path to use feels like it's more about how workspaces work, and less about how string substitution works)

Reply:
I'm not sure there's a good way to disentangle the choice of mountpath for each case (step / sidecar / shared) from the string substitution logic.  Particularly since the variables being replaced are order-dependent (we have to replace the isolated workspace mountpaths before those that are shared). I can't see an approach that doesn't either leak some workspace logic into the string substitution package or leak some string substitution logic into the workspace package.

We can always revisit?  There's nothing user-facing here that will prevent a refactor if we figure out a better separation.